src/query-builder/QueryBuilder.ts

```diff
@@ -993,51 +993,68 @@ export abstract class QueryBuilder<Entity> {
 
             if (metadata.hasRelationWithPropertyPath(path)) {
                 const relation = metadata.findRelationWithPropertyPath(path)!;
+                const isEntityTarget = relation.inverseEntityMetadata.target === entity[key].constructor;
 
-                // There's also cases where we don't want to return back all of the properties.
-                // These handles the situation where someone passes the model & we don't need to make
-                // a HUGE `where` to uniquely look up the entity.
+                const primaryColumns = relation.inverseEntityMetadata.primaryColumns;
+                const seenPrimaryColumns = new Set<string>();
+                for (const primaryColumn of primaryColumns) {
+                    const entityValueMap = primaryColumn.getEntityValueMap(entity[key]);
+                    for (const [columnName, value] of Object.entries(entityValueMap || {})) {
+                        if (value) {
+                            seenPrimaryColumns.add(columnName);
+                        } else {
+                            break;
+                        }
+                    }
+                }
+                const hasOnlyPrimaryKeys = primaryColumns.length > 0 && seenPrimaryColumns.size === Object.keys(entity[key]).length;
+
+                if (isEntityTarget || hasOnlyPrimaryKeys) {
+                    // There's also cases where we don't want to return back all of the properties.
+                    // These handles the situation where someone passes the model & we don't need to make
+                    // a HUGE `where` to uniquely look up the entity.
+
+                    // In the case of a *-to-one, there's only ever one possible entity on the other side
+                    // so if the join columns are all defined we can return just the relation itself
+                    // because it will fetch only the join columns and do the lookup.
+                    if (relation.relationType === "one-to-one" || relation.relationType === "many-to-one") {
+                        const joinColumns = relation.joinColumns
+                            .map(j => j.referencedColumn)
+                            .filter((j): j is ColumnMetadata => !!j);
+
+                        const hasAllJoinColumns = joinColumns.length > 0 && joinColumns.every(
+                            column => column.getEntityValue(entity[key], false)
+                        );
+
+                        if (hasAllJoinColumns) {
+                            paths.push(path);
+                            continue;
+                        }
+                    }
+
+                    if (relation.relationType === "one-to-many" || relation.relationType === "many-to-many") {
+                        throw new Error(`Cannot query across ${relation.relationType} for property ${path}`);
+                    }
```

Comment:
Pretty sure this needs to remain outside the `if (isEntityTarget || hasOnlyPrimaryKeys) { ... }` block as it's unrelated to the shortcuts taken.  We just do not currently support this kind of query at all.